src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -236,27 +243,103 @@ public bool RefreshOption(OptionKey2 optionKey, object? newValue)
             return true;
         }
 
+        public void AddOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+        {
+            lock (_gate)
+            {
+                if (_keepAliveTable.TryGetValue(target, out var existingHandler))
+                {
+                    var newHandler = existingHandler + handler;
+#if NET6_0_OR_GREATER
+                    _keepAliveTable.AddOrUpdate(target, newHandler);
+#else
+                    _keepAliveTable.Remove(target);
+                    _keepAliveTable.Add(target, newHandler);
+#endif
+                }
+            }
+
+            ImmutableInterlocked.Update(
+                ref _weakHandlers,
+                static (weakHandlers, handler) =>
+                {
+                    return weakHandlers
+                        .RemoveAll(WeakReferenceExtensions.IsNull)
+                        .Add(new WeakReference<EventHandler<OptionChangedEventArgs>>(handler));
+                },
+                handler);
```

Comment:
seems odd to need ImmutableInterlocked *and* a `_gate`.  Can we just do all things atomically under the one case?

Reply:
The common case is reading from this list. The code is structured so the uncommon case is the only one that needs a lock.

Reply:
➡️ Moved to #68062